vslib/SwitchPoE.cpp

```diff
@@ -0,0 +1,346 @@
+{
+    SWSS_LOG_ENTER();
+
+    if (object_type == SAI_OBJECT_TYPE_POE_DEVICE)
+    {
+        sai_object_id_t object_id;
+        sai_deserialize_object_id(serializedObjectId, object_id);
+        return createPoeDevice(object_id, switch_id, attr_count, attr_list);
+    }
+
+    if (object_type == SAI_OBJECT_TYPE_POE_PORT)
+    {
+        sai_object_id_t object_id;
+        sai_deserialize_object_id(serializedObjectId, object_id);
+        return createPoePort(object_id, switch_id, attr_count, attr_list);
+    }
+
+    return create_internal(object_type, serializedObjectId, switch_id, attr_count, attr_list);
+}
+
+sai_status_t SwitchPoE::createPoeDevice(
+        _In_ sai_object_id_t object_id,
+        _In_ sai_object_id_t switch_id,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list)
+{
+    SWSS_LOG_ENTER();
+
+    auto sid = sai_serialize_object_id(object_id);
+
+    auto power_limit = sai_metadata_get_attr_by_id(SAI_POE_DEVICE_ATTR_POWER_LIMIT_MODE, attr_count, attr_list);
+
+    CHECK_STATUS(create_internal(SAI_OBJECT_TYPE_POE_DEVICE, sid, switch_id, attr_count, attr_list));
+
+    sai_attribute_t attr;
+
+    /* if not set in create_internal then use default values*/
+    if (!power_limit)
+    {
+        attr.id = SAI_POE_DEVICE_ATTR_POWER_LIMIT_MODE;
+        attr.value.u32 = SAI_POE_DEVICE_LIMIT_MODE_CLASS;
+        CHECK_STATUS(set(SAI_OBJECT_TYPE_POE_DEVICE, object_id, &attr));
+    }
+
+    /* update the list of all POE devices on switch */
+    m_poe_device_list.push_back(object_id);
+
+    auto device_count = (uint32_t)m_poe_device_list.size();
+    attr.id = SAI_SWITCH_ATTR_POE_DEVICE_LIST;
+    attr.value.objlist.count = device_count;
+    attr.value.objlist.list = m_poe_device_list.data();
+    CHECK_STATUS(set(SAI_OBJECT_TYPE_SWITCH, switch_id, &attr));
+
+    return SAI_STATUS_SUCCESS;
+}
+
+sai_status_t SwitchPoE::createPoePort(
+        _In_ sai_object_id_t object_id,
+        _In_ sai_object_id_t switch_id,
+        _In_ uint32_t attr_count,
+        _In_ const sai_attribute_t *attr_list)
+{
+    SWSS_LOG_ENTER();
+
+    auto sid = sai_serialize_object_id(object_id);
+
+    auto power_limit = sai_metadata_get_attr_by_id(SAI_POE_PORT_ATTR_POWER_LIMIT, attr_count, attr_list);
+    auto power_priority = sai_metadata_get_attr_by_id(SAI_POE_PORT_ATTR_POWER_PRIORITY, attr_count, attr_list);
+
+    CHECK_STATUS(create_internal(SAI_OBJECT_TYPE_POE_PORT, sid, switch_id, attr_count, attr_list));
+
+    sai_attribute_t attr;
+
+    /* default admin state is down as defined in SAI */
+    attr.id = SAI_POE_PORT_ATTR_ADMIN_ENABLED_STATE;
+    attr.value.booldata = false;
+    CHECK_STATUS(set(SAI_OBJECT_TYPE_POE_PORT, object_id, &attr));
+
+    /* if not set in create_internal then use default values*/
+    if (!power_limit)
+    {
+        attr.id = SAI_POE_PORT_ATTR_POWER_LIMIT;
+        attr.value.u32 = 0;
+        CHECK_STATUS(set(SAI_OBJECT_TYPE_POE_PORT, object_id, &attr));
```

Comment:
This is just for my understanding, 

I was under the impression that this code would eventually interact with PoE controller, could you point me to code where that happens.

Reply:
This is an implementation for the virtual switch. There is no poe cotroller to communicate with, so we only report dummy data.
When vendors are writing sai for their physical devices and handling SAI_POE_* attributes then they will be communicating with the poe controller.